### PR TITLE
add all tests of boardpresentation class

### LINKED ISSUE #7: Write test cases for WoodokuBoardRepresentation

### PR DESCRIPTION
**Pull request by YANLING-H**
__validate will be tested together with WoodokuBoard

### CODE REVIEW

**Comment by Xiaorui-Huang** [comment]
Remember to set the Assignee, label if suitable, project and milestone and issue

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 1]
```diff
@@ -25,6 +25,9 @@ class _WoodokuBoardRepresentation:
     def __init__(self):
         self.__board = np.full((N, N), False)
 
+    def board(self):
+        return self.__board
+
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py:30]: Exposing a private variable is never considered as a good practice. Since this is a data class that hides the underlining implementation, `__board` should not be exposed in any case.

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import pytest
+
+from Exceptions.Exceptions import ShapeOutOfBoardError
+from WoodokuBoard import _WoodokuBoardRepresentation
+import numpy as np
+
+N = 9
+
+
+class TestWoodokuRepresentation:
```
Comment (EdwardHaoranLee) [inline Woodoku/tests/Entity/test_woodoku_board.py]: For all the test cases here, I would say it is better to not using the underlining `__board` object. Instead, just test with the method calls `add_blocks`, `remove_blocks`, `is_occupied` and `is_not_occupied`. 

For example, `test_add_blocks_add_one_block` is testing if a block is actually added after calling `add_blocks`. This could be written as 

```python
    def test_add_blocks_add_one_block(self):
        board = _WoodokuBoardRepresentation()
        board.add_blocks(self.one_block_lst)
        assert board.is_occupied(self.one_block_lst)
```

You might claim that in this above way, we are not testing `add_blocks` method alone. In fact, we do not need to. The only reason we call `add_blocks` is because we hope to get a `true` value when we call `is_occupied` later. Thus, as long as these two methods have consistent result, it should be acceptable. 

Same logic applies to `remove_blocks` and `is_not_occupied`.

I might also need to borrow some input from @Xiaorui-Huang . What do you think?

Feel free to initiate a discussion!

**Review by EdwardHaoranLee — DISMISSED** [review]

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    three_block_lst = [(1, 3), (5, 8), (0, 7)]
+    five_block_lst = three_block_lst + [(2, 1), (0, 7)]
+
+    def board_after_adding(self, rep: _WoodokuBoardRepresentation, lst: list):
+        rep.add_blocks(lst)
+        return rep.board()
+
+    def board_after_removing(self, rep: _WoodokuBoardRepresentation, lst: list):
+        rep.remove_blocks(lst)
+        return rep.board()
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]: These are set up for the tests
refer to this guide https://docs.pytest.org/en/6.2.x/xunit_setup.html or just search pytest setup or teardown 

I'm not sure about teardown, but perhaps investigate if any variables maybe persistance and not go out of scope during a test. any lingering variables should be deleted (I think either set to null or use the `del` keyword)

Another useful guide https://realpython.com/pytest-python-testing/#fixtures-managing-state-and-dependencies
(Which I haven't read all of it yet😅)

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 2 above)
Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: Since I only expose a getter instead of a direct setter,  I thought that it won't cause a large problem, but I will test methods in representation pairwisely if both of you agree.

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 3 above)
Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: I checked the same doc yesterday, but I found that `board_after_adding` and `board_after_removing` are helper functions instead of the code necessary to run before each test. My initial intention is that since both methods try to get access to the private field, I would rather create a wrapper function for the getter and call it instead of calling getters everywhere in tests.

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 2 above)
Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: Well, I'd sort of agree. The actual spirit should be to isolate each and every test and not use methods of your own to facilitate. but that is just too much setup code to justify bothering. Here is my stance reiterated https://softwareengineering.stackexchange.com/a/330314

However, this relies on the test cases being more thorough otherwise there may be trivial test errors being missed as your own setup method begins to change. 

I guess just find a balance between productive and correctness

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
+from typing import Tuple
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]: There are two approaches to balancing productivity and being thorough when testing. 
One is to test as much as possible and the other is to only test the core functionality.

I'd say that we should only test public methods and public classes, unless the private classes are complex enough to warrant a test. That means only testing Woodoku board and not the representation. But that's just a strategy to save time working on actually implementing features instead of always writing tests. 

Now, to address some issues with the tests that are already written. I'd say there need to be more cases. Empty, 1, 3, 5 blocks are just the basics, of course, you should include them, but the important tests are the edge cases. Like inserting/removing/checking invalid blocks. This should apply to all future test written that keep in mind some common errors

second note: perhaps the name of this file should be _test_WoodokuBoard.py_?

Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: We might need to reach an agreement before adding more tests to TestWoodokuRepresentation. Since the representation is a private data class, the private field will be accessed when testing each method isolated. 

There are three plans summarized according to the conversations above:

1. remove the current tests and test each method pairwisely with basics and more examples
2. keep the current tests and test more edge cases with the public getter of the private field `__board` in the representation class
3. leave the current tests as it is now and focus more on testing the public tests (ie: all other classes except `WoodokuBoardRepresentation`)

I am fine with any of the plans. You guys can vote.

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 3 above)
Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: First, the feature code is what is important. so feature code should not be altered for the convenience of the test code. 

Second, read the guide, there are methods setups/teardown that run once depending on the scenario. It could be once for the whole test file, once for the class, once for the class method or once for a function.

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 4 above)
Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: Also, I found that the module name should be short, and all-lowercase, according to [doc](https://peps.python.org/pep-0008/#package-and-module-names). Then should we keep camelcase names in src/Entity or should we change the name of test files? I am fine with either way.

Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: > 1. pairwisely

what does this mean?

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 2 above)
Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py:12]: > 
> ```python
>     def test_add_blocks_add_one_block(self):
>         board = _WoodokuBoardRepresentation()
>         board.add_blocks(self.one_block_lst)
>         assert board.is_occupied(self.one_block_lst)
> ```
> 
> You might claim that in this above way, we are not testing `add_blocks` method alone. In fact, we do not need to. The only reason we call `add_blocks` is because we hope to get a `true` value when we call `is_occupied` later. Thus, as long as these two methods have consistent result, it should be acceptable.
> 
> Same logic applies to `remove_blocks` and `is_not_occupied`.
Similarly with this example, I might combine two methods and test their combination, eg: `add_block` + `is_occupied`/`is_not_occupied` and `remove_block` + 'is_occupied`/'is_not_occupied'

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 4 above)
Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: > 
> ```python
>     def test_add_blocks_add_one_block(self):
>         board = _WoodokuBoardRepresentation()
>         board.add_blocks(self.one_block_lst)
>         assert board.is_occupied(self.one_block_lst)
> ```
> 
> You might claim that in this above way, we are not testing `add_blocks` method alone. In fact, we do not need to. The only reason we call `add_blocks` is because we hope to get a `true` value when we call `is_occupied` later. Thus, as long as these two methods have consistent result, it should be acceptable.
> 
> Same logic applies to `remove_blocks` and `is_not_occupied`.


Similarly with the example above, I might test combination such as`add_block` + `is_occupied`/`is_not_occupied` and `remove_block` + `is_occupied/is_not_occupied`

Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: > Also, I found that the module name should be short, and all-lowercase, according to [doc](https://peps.python.org/pep-0008/#package-and-module-names). Then should we keep camelcase names in src/Entity or should we change the name of test files? I am fine with either way.

refered in new issue #30. discussion should be moved there

Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: Well, I choose 1 + 2 + 3 hahah...

refactor the test written, add edge cases, and focus on public methods from now on, use your own judgement when using multiple non-tested methods in the same test function in balancing productivity and catching errors.

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 2 above)
Comment (EdwardHaoranLee) [reply Woodoku/tests/Entity/test_woodoku_board.py:12]: Yeah, I agree using one's own method when testing might not be a great idea. However, the method `board` was created just for testing. It might break the encapsulation principle so other class will also be able to access the inner data object. 

I am proposing to just use the private variable `__board` in test cases, so we could prevent that other non-testing class would have access to it. Thoughts? @YANLING-H @Xiaorui-Huang

**Comment by YANLING-H** [comment]
Since I rebase this test branch with the develop branch, there are more commits not belonging to this pr appeared, and "file changed" reflects more changes.

**Comment by Xiaorui-Huang** [comment]
> Since I rebase this test branch with the develop branch, there are more commits not belonging to this pr appeared, and "file changed" reflects more changes.

You typically force push after rebase, since rebasing changes the commit hash, so if you just push you'd be duplicating your own commit. I suggest squash and merge for this PR so that it appears as one commit (you can do that at merge time)
image
by clicking on the drop down arrow

OR if you trust your rebase skills, you can use `git rebase head~<number of commit you want to edit> -i`, but be very careful. I'd watch a tutorial before I do that. And if anything goes wrong you can always `git rebase --abort`

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 5]
```diff
@@ -1,16 +1,15 @@
...
 
 N = 9
 
 
 class TestWoodokuRepresentation:
-    @classmethod
-    def setup(cls):
-        cls.one_block_lst = [(0, 0)]
-        cls.three_block_lst = [(1, 3), (5, 8), (0, 7)]
-        cls.five_block_lst = cls.three_block_lst + [(2, 1), (0, 8)]
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]: Why did you remove the set up, can you not call the block list in @pytest.mark.parametrize ?

Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: I found that variables associated with the class such as cls.one_block are not in the same scope of @pytest.mark.parametrize

Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: I also experiment with @pytest.fixture. It doesn't work with @pytest.mark.parametrize

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 6]
```diff
@@ -20,37 +19,31 @@ def board_after_removing(self, rep: _WoodokuBoardRepresentation, lst: list):
...
         return rep._WoodokuBoardRepresentation__board
 
-    def test_add_blocks_add_one_block(self):
-        board = self.board_after_adding(_WoodokuBoardRepresentation(), self.one_block_lst)
+    @pytest.mark.parametrize("lst", [
+        one_block_lst,
+        three_block_lst,
+        five_block_lst,
+    ])
+    def test_add_blocks_add_list_of_blocks(self, lst):
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]: add type hints

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 7]
```diff
@@ -1,16 +1,15 @@
 import numpy as np
+import pytest
 
-from WoodokuBoard import _WoodokuBoardRepresentation
+from woodoku_board import _WoodokuBoardRepresentation
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]: remember it's entity.woodoku_board

**Review by Xiaorui-Huang — CHANGES_REQUESTED** [review]
edit the changes listed in the conversations

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
+import numpy as np
+import pytest
+from numpy import ndarray
+
+from woodoku_board import _WoodokuBoardRepresentation
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]:
```suggestion
from entity.woodoku_board import _WoodokuBoardRepresentation
```

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
+import numpy as np
```
Comment (Xiaorui-Huang) [inline Woodoku/tests/Entity/test_woodoku_board.py]: file in the wrong folder

Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: What do you mean?

Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: **
image
**
file in the wrong folder

**Review by Xiaorui-Huang — DISMISSED** [review]

**Review by Xiaorui-Huang — APPROVED** [review]
